File: src/main/java/com/example/travelator/optional/java/advanced1/Legs.java
package com.example.travelator.optional.java.advanced1;

import com.example.travelator.optional.java.Leg;

import java.time.Duration;
import java.util.Comparator;
import java.util.List;
import java.util.Optional;

public class Legs {
	
	private Legs() {
	}
	
	public static Optional<Leg> findLongestLegOver(List<Leg> legs, Duration duration) {
		return legs.stream()
			.max(Comparator.comparing(Leg::getPlannedDuration))
			.filter(leg -> isLongerThan(leg, duration));
	}
	
	private static boolean isLongerThan(Leg leg, Duration duration) {
		return leg.getPlannedDuration().compareTo(duration) > 0;
	}
}

File: src/main/java/com/example/travelator/optional/advanced1/Legs.kt
package com.example.travelator.optional.advanced1

import com.example.travelator.optional.kotlin.Leg
import kotlin.time.Duration

/**
 * 최상위 함수
 */
fun longestLegOver(
    legs: List<Leg>,
    duration: Duration,
): Leg? {
    var result: Leg? = null
    for (leg in legs) {
        if (leg.isLongerThan(duration)) {
            if (result == null || leg.isLongerThan(result.plannedDuration)) {
                result = leg
            }
        }
    }
    return result
}

/**
 * Convert parameter to receiver(파라미터를 수신 객체로 변환)
 */
private fun Leg.isLongerThan(duration: Duration) =
    plannedDuration > duration